src/libpmempool/replica.c

```diff
@@ -528,29 +626,201 @@ check_checksums_and_signatures(struct pool_set *set,
 			if (!util_checksum(hdr, sizeof(*hdr), &hdr->checksum, 0,
 					POOL_HDR_CSUM_END_OFF)) {
 				ERR("invalid checksum of pool header");
-				rep_hs->part[p] |= IS_BROKEN;
+				rep_hs->part[p].flags |= IS_BROKEN;
 			} else if (util_is_zeroed(hdr, sizeof(*hdr))) {
-					rep_hs->part[p] |= IS_BROKEN;
+					rep_hs->part[p].flags |= IS_BROKEN;
 			}
 
 			enum pool_type type = pool_hdr_get_type(hdr);
 			if (type == POOL_TYPE_UNKNOWN) {
 				ERR("invalid signature");
-				rep_hs->part[p] |= IS_BROKEN;
+				rep_hs->part[p].flags |= IS_BROKEN;
 			}
 		}
 	}
 	return 0;
 }
 
 /*
- * check_badblocks -- (internal) check if replica contains bad blocks
+ * replica_badblocks_recovery_file_save -- save bad blocks in the bad blocks
+ *                                         recovery file before clearing them
+ */
+static int
+replica_badblocks_recovery_file_save(struct part_health_status *part_hs)
+{
+	LOG(3, "part_health_status %p", part_hs);
+
+	ASSERTeq(part_hs->exist_recovery_file, 1);
+	ASSERTne(part_hs->recovery_file, NULL);
+
+	struct badblocks *bbs = &part_hs->bbs;
+	char *path = part_hs->recovery_file;
+	int ret = 0;
+
+	int fd = os_open(path, O_WRONLY);
+	if (fd < 0) {
+		ERR("!opening bad blocks recovery file failed -- '%s'", path);
+		return -1;
+	}
+
+	FILE *recovery_file = os_fdopen(fd, "w");
+	if (recovery_file == NULL) {
+		ERR(
+			"!opening a file stream for bad blocks recovery file failed -- '%s'",
+			path);
+		os_close(fd);
+		return -1;
+	}
+
+	for (unsigned i = 0; i < bbs->bb_cnt; i++) {
+		fprintf(recovery_file, "%llu %u\n",
+			bbs->bbv[i].offset, bbs->bbv[i].length);
+	}
+
+	/* save the finish flag */
+	fprintf(recovery_file, "0 0\n");
```

Comment:
should we maybe flush & fsync once after writing the badblocks and then the second time after writing the finish flag?

Reply:
Why? Without the finish flag, the recovery file is not valid.